### PR TITLE
[Data] [no_early_kickoff] Add `column` API to Dataset

### LINKED ISSUE #29018: [Datasets] Add `Dataset.columns`
### Description

Currently, I do this

```
>>> ds.schema().names
['spam', 'ham']
```

But, I'd like to do this

```
>>> ds.columns
['spam', 'ham']
```

A potential signature is:

```
@property
def columns(self) -> list[str]:
    """The column names.

     Raises:
          NotImplementedError: if this dataset isn't tabular.
     """
     ...
```

### Use case

QOL

### PR DESCRIPTION
**Pull request by amogkam**
Adds columns API to Dataset to be able to see the columns of the Dataset.

Closes https://github.com/ray-project/ray/issues/29018

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_consumption.py [hunk 1]
```diff
@@ -249,6 +249,12 @@ def test_schema_lazy(ray_start_regular_shared):
     assert ds._plan.execute()._num_computed() == 0
 
 
+def test_columns(ray_start_regular_shared):
+    ds = ray.data.range(100, parallelism=10)
```
Comment (bveeramani) [inline python/ray/data/tests/test_consumption.py:253]: Nit: Any reason we need the `parallelism=10`?
```suggestion
    ds = ray.data.range(1)
```

Writing minimally passing tests is a best practice.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -4218,6 +4218,12 @@ def _aggregate_result(self, result: Union[Tuple, Mapping]) -> U:
         else:
             return result
 
+    @PublicAPI(stability="alpha")
+    @property
```
Comment (bveeramani) [inline python/ray/data/dataset.py:4222]: Users expect properties to return immediately. So, if `self.schema()` is slow, users might be surprised.

That said, I think it's reasonable to make `columns` a property. No strong opinion either way.

**Review by bveeramani — APPROVED** [review]
LGTM

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -4218,6 +4218,12 @@ def _aggregate_result(self, result: Union[Tuple, Mapping]) -> U:
         else:
             return result
 
+    @PublicAPI(stability="alpha")
```
Comment (ericl) [inline python/ray/data/dataset.py]:
```suggestion
```

This isn't necessary.

File: python/ray/data/dataset.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/dataset.py]: updated to a method...we also need to accept `fetch_if_missing`

File: python/ray/data/dataset.py (hunk 3 above)
Comment (amogkam) [reply python/ray/data/dataset.py]: removed

**Comment by ericl** [comment]
Tests/doctests are broken.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (bveeramani) [reply python/ray/data/dataset.py]: For my own understanding, why don't we need this? We use `@PublicAPI(stability="alpha")` for other `Dataset` methods, and I'm not sure how else we'd distinguish that this particular method is unstable.

Comment (ericl) [reply python/ray/data/dataset.py]: I think this is simple enough we can immediately graduate this to Public.

**Comment by amogkam** [comment]
failing tests are unrelated, going to merge

**Pull request merged**